src/trio/_core/_ki.py

```diff
@@ -117,90 +201,37 @@ def currently_ki_protected() -> bool:
-        # In some version of Python, isgeneratorfunction returns true for
-        # coroutine functions, so we have to check for coroutine functions
-        # first.
-        if inspect.iscoroutinefunction(fn):
-
-            @wraps(fn)
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:  # type: ignore[misc]
-                # See the comment for regular generators below
-                coro = fn(*args, **kwargs)
-                assert coro.cr_frame is not None, "Coroutine frame should exist"
-                coro.cr_frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return coro  # type: ignore[return-value]
-
-            return wrapper
-        elif inspect.isgeneratorfunction(fn):
-
-            @wraps(fn)
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:  # type: ignore[misc]
-                # It's important that we inject this directly into the
-                # generator's locals, as opposed to setting it here and then
-                # doing 'yield from'. The reason is, if a generator is
-                # throw()n into, then it may magically pop to the top of the
-                # stack. And @contextmanager generators in particular are a
-                # case where we often want KI protection, and which are often
-                # thrown into! See:
-                #     https://bugs.python.org/issue29590
-                gen = fn(*args, **kwargs)
-                gen.gi_frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return gen  # type: ignore[return-value]
-
-            return wrapper
-        elif inspect.isasyncgenfunction(fn) or legacy_isasyncgenfunction(fn):
-
-            @wraps(fn)  # type: ignore[arg-type]
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:  # type: ignore[misc]
-                # See the comment for regular generators above
-                agen = fn(*args, **kwargs)
-                agen.ag_frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return agen  # type: ignore[return-value]
-
-            return wrapper
-        else:
-
-            @wraps(fn)
-            def wrapper(*args: ArgsT.args, **kwargs: ArgsT.kwargs) -> RetT:
-                sys._getframe().f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED] = enabled
-                return fn(*args, **kwargs)
+_T_supports_code = TypeVar("_T_supports_code", bound=_SupportsCode)
 
-            return wrapper
 
-    return decorator
+def enable_ki_protection(f: _T_supports_code, /) -> _T_supports_code:
+    """Decorator to enable KI protection."""
+    orig = f
 
+    if legacy_isasyncgenfunction(f):
+        f = f.__wrapped__  # type: ignore
 
-# pyright workaround: https://github.com/microsoft/pyright/issues/5866
-class KIProtectionSignature(Protocol):
-    __name__: str
+    code = f.__code__.replace()
+    _CODE_KI_PROTECTION_STATUS_WMAP[code] = True
+    f.__code__ = code
```

Comment:
Why do you need to generate a new code object here rather than adding `f.__code__` to the map?

Reply:
This is required for test_ki_does_not_leak_accross_different_calls_to_inner_functions to pass

Reply:
This seems like a very big hammer to serve a pretty obscure case. For top-level functions, it's minorly increasing import time to no benefit. For closures, each instance of the closure now has to carry around its own 200-byte code object, instead of all of them being able to share one. How would you feel about instead documenting that we expect `@enable/disable_ki_protection` to be applied or not to a given function in the source file in a consistent way, and optionally adding another interface that does dup the code object for use when the protection status might be runtime-dependent?

Reply:
I think there's other cases where functions can get the same code object, like importlib.reload

I think by ignoring this hang-nail we can use a regular equality based WeakKeyDictionary rather than building a complicated identity based one

Reply:
How about we issue an exception if the code object is already in the weak key dictionary or if the function doesn't have an importable name?

Reply:
I still would prefer to just not worry about enforcing this (though it's good to document the caveat and provide an alternative that fully supports this sort of runtime variation). All the possible cures seem worse than the disease to me. `@disable_ki_protection` is basically never used, so the practical failure mode is that a conditionally-KI-protected closure will be always KI protected instead. It's really hard for me to imagine any actual harm coming from that. Is there a specific scenario you're thinking of that makes you more concerned about this case?

I would be OK with a warning in the cases you describe, as long as there's something reasonable the user can do to satisfy it (like a kwarg for the decorator that enables dup'ing the code object or that promises all the registrations will match). I don't think it should be a hard error, because this would also fire for a closure that is consistently KI-protected every time it's defined.

Reply:
hmm I think I'll remove the code object copy for now